pkg/lib/ldap/client.go

```diff
@@ -0,0 +1,106 @@
+package ldap
+
+import (
+	"errors"
+	"net/url"
+
+	"github.com/go-ldap/ldap/v3"
+
+	"github.com/authgear/authgear-server/pkg/api"
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/ldaputil"
+)
+
+type Client struct {
+	Config config.LDAPServerConfig
+	conn   *ldap.Conn
+}
+
+func NewClient(config config.LDAPServerConfig) *Client {
+	return &Client{
+		Config: config,
+	}
+}
+
+func (c *Client) Connect() error {
+	url, err := url.Parse(c.Config.URL)
+	if err != nil {
+		return err
+	}
+
+	if url.Port() == "" {
+		if url.Scheme == "ldaps" {
+			c.Config.URL = c.Config.URL + ":636"
```

Comment:
You modify something someone pass to you. **Never do this**.

Instead, you should modify something you create in this function, that is, `url`.
Another problem, you name the variable `url`, which clashes with the package `url`. That would make you unable to reference to `url` (the package). So, please name it `u`.

To summaries, What you need to do here is 

```
u, err := url.Parse(c.Config.URL)
if err != nil {
  return err
}

if u.Port() == "" {
  switch u.Scheme {
  case "ldaps":
    u.Host = u.Host + ":636"
  case "ldap":
    u.Host = u.Host + ":389"
  }
}

ldapURLString := u.String()
conn, err := ldap.DialURL(ldapURLString)
```